### PR TITLE
update subject info definition according to sub_ids

### PR DESCRIPTION
**Pull request by fimbault**
Reflect the various scopes possible in sub_ids #75 #198
Some are local to the AS, some enable dereferencing

### CODE REVIEW

**Comment by yaronf** [comment]
I'm confused. This reflects that we want to support both locally-scoped subjects and also non-locally ("federated"?) scoped ones. So does the comment in #75 with `iss_sub` subject identifiers. But the move to `opaque` means that we expect the AS to identify a subject using a single string. Why are we getting rid of the issuer URI?

**Comment by fimbault** [comment]
We're not getting rid of iss_sub, which is also a valid sub_id.

**Comment by jricher** [comment]
I don't think I agree with this change -- even in the case of an AS sending a globally-scoped identifier like iss/sub (or email, or phone, etc), the _connection_ between the user and that identifier is still being asserted locally by the AS. After all, nothing stops a rogue AS from assering an iss/sub pair that it is not the issuer for. And in the case of a user logging in to the AS with a federated identity, that might actually be desirable to know. 

The intent of the previous language was to say that the AS is only its own authority and not able to make truly global trusted statements. I think that this can be said better but I don't think removing "locally" gets at that and leaves it too open for misinterpretation.

**Comment by fimbault** [comment]
The rationale is that "asserted by the AS" already provides that information, while the scope of the identifier is defined by the choice of sub_id.

**Comment by agropper** [comment]
I'm always confused by the meaning of subject identifiers in authorization
protocols, not just GNAP. I wonder if this
https://mailarchive.ietf.org/arch/msg/txauth/ytSC3a_bw6LicwjU2YaBZpkS7z4/
framing
would be useful?

**Comment by fimbault** [comment]
Add some text in the spec saying that the assertion is made locally by the AS

**Review by jricher — APPROVED** [review]

File: draft-ietf-gnap-core-protocol.md [hunk 1]
```diff
@@ -2256,7 +2256,7 @@ Subject identifiers returned by the AS SHOULD uniquely identify the RO at the
 AS. Some forms of subject identifier are opaque to the client instance (such as the subject of an 
 issuer and subject pair), while others forms (such as email address and phone number) are
```
Comment (yaronf) [inline draft-ietf-gnap-core-protocol.md:2257]: while others -> while other

File: draft-ietf-gnap-core-protocol.md [hunk 2]
```diff
@@ -2256,7 +2256,7 @@ Subject identifiers returned by the AS SHOULD uniquely identify the RO at the
...
 intended to allow the client instance to correlate the identifier with other account information
-at the client instance. The client instance MUST NOT request or use any returned subject identifiers for communication
+at the client instance. The AS MUST validate the correspondance between the user and the returned subject identifiers. The client instance MUST NOT request or use any returned subject identifiers for communication
```
Comment (yaronf) [inline draft-ietf-gnap-core-protocol.md:2259]: correspondance - typo

"The AS MUST validate" appears to be an important security check, but it's not clear to me what we're expecting the AS to do. It seems to me that the AS, once it has authenticated the RO (a.k.a. end user) simply returns all it knows about the RO. So there's nothing to validate.

**Review by yaronf — CHANGES_REQUESTED** [review]

File: draft-ietf-gnap-core-protocol.md (hunk 2 above)
Comment (jricher) [reply draft-ietf-gnap-core-protocol.md:2259]: @yaronf the process you're describing is one way to do the validation in question (arguably the most common way).

Comment (yaronf) [reply draft-ietf-gnap-core-protocol.md:2259]: In that case I would say the proposed language is confusing. How about:

> The AS MUST ensure that the returned subject identifiers only apply to the authenticated end user.

Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:2259]:
```suggestion
at the client instance. The AS MUST ensure that the returned subject identifiers only apply to the authenticated end user. The client instance MUST NOT request or use any returned subject identifiers for communication
```

Comment (jricher) [reply draft-ietf-gnap-core-protocol.md]: Accepted Yaron's suggested text.

**Pull request merged**